File: implementations/rust/src/diagnostic.rs
use proc_macro2::Span;
use serde::Serialize;

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Error,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct Diagnostic {
    #[serde(rename = "rule_id")]
    pub rule_id: String,
    pub severity: Severity,
    pub message: String,
    pub file: String,
    pub line: usize,
    pub column: usize,
}

impl Diagnostic {
    pub fn new(
        rule_id: impl Into<String>,
        message: impl Into<String>,
        file: impl Into<String>,
        line: usize,
        column: usize,
    ) -> Self {
        Self {
            rule_id: rule_id.into(),
            severity: Severity::Error,
            message: message.into(),
            file: file.into(),
            line,
            column,
        }
    }

    pub fn from_span(
        rule_id: impl Into<String>,
        message: impl Into<String>,
        file: impl Into<String>,
        span: Span,
    ) -> Self {
        let start = span.start();
        Self::new(rule_id, message, file, start.line, start.column + 1)
    }
}

pub fn diagnostic_at_offset(
    rule_id: impl Into<String>,
    message: impl Into<String>,
    file: impl Into<String>,
    source: &str,
    offset: usize,
) -> Diagnostic {
    let (line, column) = location(source, offset);
    Diagnostic::new(rule_id, message, file, line, column)
}

fn location(source: &str, offset: usize) -> (usize, usize) {
    let bounded = offset.min(source.len());
    let mut line = 1;
    let mut column = 1;

    for (index, ch) in source.char_indices() {
        if index >= bounded {
            break;
        }
        if ch == '\n' {
            line += 1;
            column = 1;
        } else {
            column += 1;
        }
    }

    (line, column)
}

File: implementations/typescript/src/diagnostic.ts
export type Severity = "error";

export interface Diagnostic {
  rule_id: string;
  severity: Severity;
  message: string;
  file: string;
  line: number;
  column: number;
}

export function diagnostic(
  ruleId: string,
  message: string,
  file: string,
  line: number,
  column: number,
): Diagnostic {
  return {
    rule_id: ruleId,
    severity: "error",
    message,
    file,
    line,
    column,
  };
}

export function sortDiagnostics(items: Diagnostic[]): Diagnostic[] {
  return [...items].sort((left, right) => {
    if (left.file !== right.file) {
      return left.file < right.file ? -1 : 1;
    }
    if (left.line !== right.line) {
      return left.line - right.line;
    }
    if (left.column !== right.column) {
      return left.column - right.column;
    }
    if (left.rule_id !== right.rule_id) {
      return left.rule_id < right.rule_id ? -1 : 1;
    }
    return 0;
  });
}

export function renderText(items: Diagnostic[]): string {
  if (items.length === 0) {
    return "";
  }
  const first = items[0];
  return `${first.file}:${first.line}:${first.column}: ${first.rule_id}: ${first.message}\n`;
}

export function renderJSON(items: Diagnostic[]): string {
  return `${JSON.stringify({ diagnostics: items }, null, 2)}\n`;
}
